MAINT: Use yarn as package manager

stevejpurves opened the pull request:
Using `yarn` in place of `npm` as the package manager

 - tested with `yarn` and resolved a number of issues around package dependencies.
 - updated contributing docs
 - added advisory around use of v7 and possible peer dependency errors
 - updates the `docs` build
 - updated `ci` actions
 - updates `dependabot`

Addresses #426

stevejpurves commented:
Leaving this marked as WIP until ci goes green.

stevejpurves commented:
ok, this is green and everything seems happy. I think this should make our CI checks more reliable 👍  and would love to get this in.

NOTE: if we accept this change, we should just delete all open dependabot PRs and wait on new ones coming in. Existing dependabot PRs are trying to patch `package-lock.json` which is now gone and we have updated dependencies with the PR anyways, because of the new `yarn-lock.json`

moorepants commented:
Got these warnings on first install:

```
warning "@jupyter-widgets/base > @lumino/coreutils@1.8.0" has unmet peer dependency "crypto@1.0.1".
warning "@jupyterlab/theme-light-extension > @jupyterlab/application > @jupyterlab/ui-components@3.0.7" has unmet peer dependency "react@^17.0.1".
warning "@jupyterlab/testutils > ts-jest@25.5.1" has unmet peer dependency "typescript@>=3.4 <4.0".
warning " > jest-puppeteer@4.4.0" has incorrect peer dependency "puppeteer@>= 1.5.0 < 3".
warning " > ts-jest@27.0.4" has incorrect peer dependency "jest@^27.0.0".
warning " > ts-jest@27.0.4" has unmet peer dependency "typescript@>=3.8 <5.0".
```

moorepants commented on .github/workflows/test.yml:
```diff
@@ -31,19 +31,19 @@ jobs:
+          yarn install --frozen-lockfile
 
-      - name: "`npm run fmt` and check for changes"
+      - name: "`nary run fmt` and check for changes"
```
```suggestion
      - name: "`yarn run fmt` and check for changes"
```

moorepants commented:
When I run `yarn run test` locally jest isn't found. Should `yarn install` add jest? Or is there a separate lock file for dev dependencies?

choldgraf commented:
Thanks for this PR - we should really get input from @minrk to make sure he is OK with this. (or @minrk feel free to say that you are fine with whatever @stevejpurves is proposing if you don't have the bandwidth to review etc!)

moorepants commented on docs/contribute.md:
```diff
@@ -26,39 +26,63 @@ In order to get Thebe running locally, you'll need to have Node installed on you
+
+#### Peer Dependency Errors
+
+Due to recent changes in `npm` as of v7 peer dependency issues are flagged as critical errors. However, may projects have not yet resolves these issues in their code base, this means that a project's co-dependencies can prevent installation of a package. This is a know issue, to resolve this there are two options:
```
```suggestion
Due to recent changes in `npm` as of v7 peer dependency issues are flagged as critical errors. However, many projects have not yet resolved these issues in their code base, this means that a project's co-dependencies can prevent installation of a package. This is a know issue, to resolve this there are two options:
```

moorepants commented:
There is an empty `yarn-lock.json` file committed along with the two `yarn.lock` files. Maybe that is unintentional?

moorepants commented:
This looks good to me so far.

1. I'm not sure how to install the full development dependencies to run the tests.
2. We'll likely have to check the production build commands on master when we do a release for a full check of the CI.

Thanks @stevejpurves !

stevejpurves commented:
@moorepants to your points above

 - yarn-lock.json removed and the docs makefile is now fixed
 - I have addressed a couple of them. I'm tempted to say leave the others, though? (react, typescript) these are our dependencies telling us we are responsible for providing these peer dependencies, but looking at each warning there, I don't think we are using our dependencies in way where we need to install react, typescript, so would be best not to add them.

moorepants commented:
I'll try again.

moorepants commented:
Everything is working now. Not sure what I did wrong before, maybe I didn't delete the `node_modules` directory first.

moorepants commented:
+1 from me on this for merging.

choldgraf reviewed (approved):
I also think this is a nice enhancement! Don't mistake my comments above for a lack of enthusiasm!

stevejpurves commented:
no worries @choldgraf it's a good point that we're all on different cadences here and need to allow some time for wider contributions :)

stevejpurves commented:
Are we good to merge this one folks? @moorepants @choldgraf (I have no merge button :) )

choldgraf commented:
Sounds good! Thanks for the ping 🙂
